File: clients/client-python/gravitino/dto/metadata_object_dto.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from dataclasses_json import config

from gravitino.api.metadata_object import MetadataObject
from gravitino.utils.precondition import Precondition


@dataclass
class MetadataObjectDTO(MetadataObject):
    """Represents a Metadata Object DTO (Data Transfer Object)."""

    _type: MetadataObject.Type = field(metadata=config(field_name="type"))
    _full_name: str = field(metadata=config(field_name="fullName"))

    _name: str = field(init=False, default="")
    _parent: Optional[str] = field(init=False, default=None)

    def __post_init__(self) -> None:
        self.set_full_name(self._full_name)

    @staticmethod
    def builder() -> MetadataObjectDTO.Builder:
        return MetadataObjectDTO.Builder()

    def type(self) -> MetadataObject.Type:
        return self._type

    def parent(self) -> Optional[str]:
        return self._parent

    def name(self) -> str:
        return self._name

    def set_full_name(self, full_name: str) -> None:
        """
        Sets the full name of the metadata object.

        Args:
            full_name (str): The full name of the metadata object.
        """
        index = full_name.rfind(".")
        if index == -1:
            self._parent = None
            self._name = full_name
        else:
            self._parent = full_name[:index]
            self._name = full_name[index + 1 :]

    class Builder:
        def __init__(self) -> None:
            self._full_name: str = ""
            self._type: MetadataObject.Type = None

        def full_name(self, full_name: str) -> MetadataObjectDTO.Builder:
            self._full_name = full_name
            return self

        def type(self, _type: MetadataObject.Type) -> MetadataObjectDTO.Builder:
            self._type = _type
            return self

        def build(self) -> MetadataObjectDTO:
            Precondition.check_string_not_empty(
                self._full_name, "Full name is required and cannot be empty."
            )
            Precondition.check_argument(
                self._type is not None, "Type is required and cannot be None."
            )

            return MetadataObjectDTO(
                self._type,
                self._full_name,
            )

File: common/src/main/java/org/apache/gravitino/dto/tag/MetadataObjectDTO.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.tag;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.EqualsAndHashCode;
import org.apache.gravitino.MetadataObject;

/** Represents a Metadata Object DTO (Data Transfer Object). */
@EqualsAndHashCode
public class MetadataObjectDTO implements MetadataObject {

  private String parent;

  private String name;

  @JsonProperty("type")
  private Type type;

  private MetadataObjectDTO() {}

  @Override
  public String parent() {
    return parent;
  }

  @Override
  public String name() {
    return name;
  }

  @Override
  public Type type() {
    return type;
  }

  /**
   * @return The full name of the metadata object.
   */
  @JsonProperty("fullName")
  public String getFullName() {
    return fullName();
  }

  /**
   * Sets the full name of the metadata object. Only used by Jackson deserializer.
   *
   * @param fullName The full name of the metadata object.
   */
  @JsonProperty("fullName")
  public void setFullName(String fullName) {
    int index = fullName.lastIndexOf(".");
    if (index == -1) {
      parent = null;
      name = fullName;
    } else {
      parent = fullName.substring(0, index);
      name = fullName.substring(index + 1);
    }
  }

  /**
   * @return a new builder for constructing a Metadata Object DTO.
   */
  public static Builder builder() {
    return new Builder();
  }

  /** Builder for constructing a Metadata Object DTO. */
  public static class Builder {

    private final MetadataObjectDTO metadataObjectDTO = new MetadataObjectDTO();

    /**
     * Sets the parent of the metadata object.
     *
     * @param parent The parent of the metadata object.
     * @return The builder.
     */
    public Builder withParent(String parent) {
      metadataObjectDTO.parent = parent;
      return this;
    }

    /**
     * Sets the name of the metadata object.
     *
     * @param name The name of the metadata object.
     * @return The builder.
     */
    public Builder withName(String name) {
      metadataObjectDTO.name = name;
      return this;
    }

    /**
     * Sets the type of the metadata object.
     *
     * @param type The type of the metadata object.
     * @return The builder.
     */
    public Builder withType(Type type) {
      metadataObjectDTO.type = type;
      return this;
    }

    /**
     * @return The constructed Metadata Object DTO.
     */
    public MetadataObjectDTO build() {
      return metadataObjectDTO;
    }
  }
}
